### ISSUE TITLE
The `JsonPathDetector` class does not handle empty objects correctly

### ISSUE THREAD

**Issue by tetrachromium** [cdauth/json-stream-es#1]
The `JsonPathDetector` class handles JSON objects and arrays slightly differently, due to the need to recognize an object's property keys. To handle this, the `JsonPathDetector` state machine uses an additional `"pending"` state when processing objects. During the `"pending"` state, no path item has been pushed for properties of the object.

When the closing brace of an _empty_ object is encountered, the state is still `"pending"` since no properties have been seen. The current implementation of the state machine currently assumes that a path item has been pushed. It therefore incorrectly pops the key of the _enclosing_ object or array.

Consider the following JSON fragment:

```
{
  "outer": {
    "inner": {
    } // Should have path [ "outer", "inner" ], but has path [ "outer" ]
  }
}
```

When a `JsonPathDetector` object encounters the closing brace of the `"inner"` object, the path is just `[ "outer", "inner" ]` because no nested properties have been found. Ordinarily the key of the last nested property would be popped, but the `"inner"` key is popped instead. The `JsonChunk` is therefore incorrectly given an `[ "outer" ]` path. The paths of all subsequent `JsonChunk` objects will then be incorrect.

**Comment by tetrachromium** [comment]
I have implemented a unit test and fix for this issue, but cannot push it due to insufficient access.

**Comment by tetrachromium** [comment]
Unit test that illustrates the problem:

```
test("PathDetector handles empty object", async () => {
	const stream = serializeJsonValue({
		outer: {
			inner: {}
		}
	}).pipeThrough(new JsonPathDetector());

	expect(await streamToArray(stream)).toEqual([
		{ ...objectStart(), path: [] },
		{ ...stringStart(StringRole.KEY), path: [] },
		{ ...stringChunk("outer", StringRole.KEY), path: [] },
		{ ...stringEnd(StringRole.KEY), path: [] },
		{ ...colon(), path: [] },
		{ ...objectStart(), path: ["outer"] },
		{ ...stringStart(StringRole.KEY), path: ["outer"] },
		{ ...stringChunk("inner", StringRole.KEY), path: ["outer"] },
		{ ...stringEnd(StringRole.KEY), path: ["outer"] },
		{ ...colon(), path: ["outer"] },
		{ ...objectStart(), path: ["outer", "inner"] },
		{ ...objectEnd(), path: ["outer", "inner"] },
		{ ...objectEnd(), path: ["outer"] },
		{ ...objectEnd(), path: [] }
	]);
});
```

**Comment by tetrachromium** [comment]
Pull request https://github.com/cdauth/json-stream-es/pull/2 has been created to address this issue.

**Comment by cdauth** [comment]
Brilliant, thanks for the report, the detailed explanation and the fix! I've released v1.2.1 containing the fix.
